be/src/storage/lake/rowset.cpp

```diff
@@ -102,53 +77,41 @@ StatusOr<std::vector<ChunkIteratorPtr>> Rowset::get_segment_iterators(const vect
             continue;
         }
 
-        auto res = seg_ptr->new_iterator(segment_schema, seg_options);
+        auto res = seg_ptr->new_iterator(*segment_schema, seg_options);
         if (res.status().is_end_of_file()) {
             continue;
         }
         if (!res.ok()) {
             return res.status();
         }
-        if (segment_schema.num_fields() > schema.num_fields()) {
-            tmp_seg_iters.emplace_back(vectorized::new_projection_iterator(schema, std::move(res).value()));
+        if (segment_schema != &schema) {
+            segment_iterators.emplace_back(vectorized::new_projection_iterator(schema, std::move(res).value()));
         } else {
-            tmp_seg_iters.emplace_back(std::move(res).value());
+            segment_iterators.emplace_back(std::move(res).value());
         }
     }
-
-    std::vector<ChunkIteratorPtr> segment_iterators;
-    if (tmp_seg_iters.empty()) {
-        // nothing to do
-    } else if (is_overlapped()) {
-        for (auto& iter : tmp_seg_iters) {
-            auto wrapper = std::make_shared<SegmentIteratorWrapper>(std::move(iter));
-            segment_iterators.emplace_back(std::move(wrapper));
-        }
+    if (segment_iterators.empty()) {
+        return vectorized::new_empty_iterator(*segment_schema, options.chunk_size);
+    } else if (segment_iterators.size() == 1) {
+        return segment_iterators[0];
+    } else if (options.sorted) {
+        return vectorized::new_heap_merge_iterator(segment_iterators);
```

Comment:
The reader will not be used in vertical compaction if not return segment iterators.

Reply:
I'll add a new API to get segment iterators when needed.